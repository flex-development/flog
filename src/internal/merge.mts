/**
 * @file Internal - merge
 * @module log/internal/merge
 */

import { ksort, isObjectPlain as pojo } from '@flex-development/tutils'

/**
 * Merge one or more objects into a single object.
 *
 * Source objects are applied from left to right. Subsequent sources overwrite
 * property assignments of previous sources.
 *
 * @internal
 *
 * @this {void}
 *
 * @param {Record<PropertyKey, any>} target
 *  Target object
 * @param {Record<PropertyKey, any>[]} sources
 *  Source object(s)
 * @return {Record<PropertyKey, any>}
 *  Merged object
 */
function merge(
  this: void,
  target: Record<PropertyKey, any>,
  ...sources: Record<PropertyKey, any>[]
): Record<PropertyKey, any> {
  return ksort(sources.reduce<Record<PropertyKey, any>>((acc, source) => {
    return [
      ...Object.getOwnPropertySymbols(source),
      ...Object.getOwnPropertyNames(source)
    ].reduce((target, property) => {
      /**
       * Source object value.
       *
       * @var {any} value
       */
      let value: any = source[property]

      if (Object.prototype.hasOwnProperty.call(target, property)) {
        /**
         * Target object value.
         *
         * @const {any} targetValue
         */
        const targetValue: any = target[property]

        if (pojo(targetValue) && pojo(value)) {
          value = merge(targetValue, value)
        }
      }

      return target[property] = value, target
    }, acc)
  }, target))
}

export default merge
