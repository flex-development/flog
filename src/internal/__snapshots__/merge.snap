// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:internal/merge > should return merged object (0) 1`] = `
{
  "color": "gray",
  "level": 4,
  "type": "debug",
}
`;

exports[`unit:internal/merge > should return merged object (1) 1`] = `
{
  "debug": {
    "color": "gray",
    "level": 4,
    "type": "debug",
  },
  "error": {
    "badge": true,
    "color": "red",
    "format": {
      "date": false,
    },
    "level": 0,
    "type": "error",
  },
  "fail": {
    "color": "red",
    "level": 0,
    "type": "fail",
  },
  "info": {
    "color": "cyan",
    "level": 3,
    "type": "info",
  },
  "inspect": {
    "level": 4,
    "type": "inspect",
  },
  "log": {
    "level": 2,
  },
  "start": {
    "color": "magenta",
    "level": 3,
    "type": "start",
  },
  "success": {
    "color": "green",
    "level": 3,
    "type": "success",
  },
  "warn": {
    "color": "yellow",
    "level": 1,
    "type": "warn",
  },
}
`;
